prompt2model/model_trainer/generate.py

```diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+
+            if isinstance(optuna_value, dict):
+                allowed_optuna_keys = set(
+                    optuna_hp_selector._example_hyperparameter_space.keys()
+                )
+                assert set(optuna_value.keys()) <= allowed_optuna_keys
+                optuna_hp_space_dict = optuna_value
+            else:
+                optuna_hp_space_dict = optuna_hp_selector._example_hyperparameter_space
+
+            logger.info("Starting hyperparameter search using Optuna")
+            return optuna_hp_selector.select_from_hyperparameters(
+                training_sets=training_datasets,
+                validation=validation_datasets,
+                hyperparameters=optuna_hp_space_dict,
+            )
+
+        elif "gpt" in hyperparameters:
+            raise NotImplementedError(
+                "Hyperparameter search using gpt is not yet implemented"
+            )
+
+        elif "hybrid" in hyperparameters:
+            raise NotImplementedError("Hybrid search is not yet implemented")
+
+        else:
+            raise KeyError("Invalid Key given")
+
     def train_model(
         self,
         hyperparameter_choices: dict[str, Any],
         training_datasets: list[datasets.Dataset],
         validation_datasets: list[datasets.Dataset] | None = None,
     ) -> tuple[transformers.PreTrainedModel, transformers.PreTrainedTokenizer]:
-        """Train a text generation model.
+        """Train a text generation model with automated hyperparameter search.
+
+        The hyperparameter_choices
+        parameter should follow this schema as shown in this below example.
+
+        ```python
+        hyperparameter_choice = {
+            "static_hyperparameters": {
+                "output_dir": "./result",
+                "logging_steps": 1,
+                "save_strategy": "no",
+                "num_train_epochs": 10,
+                "per_device_train_batch_size": 100,
+                "warmup_steps": 0,
+                "weight_decay": 0.01,
+                "logging_dir": "./result",
+                "learning_rate": 1e-4,
+                "evaluation_strategy": "epoch",
+                "test_size": 0.15,
+            },
+
+            "optuna": {
+                "min_num_train_epochs": 5,
+                "max_num_train_epochs": 10,
+                "save_strategy": ["epoch", "steps", "no"],
+                "evaluation_strategy": ["epoch", "no"],
+                "per_device_train_batch_size": [4, 8, 16, 32],
+                "min_weight_decay": 1e-5,
+                "max_weight_decay": 1e-1,
+                "min_learning_rate": 1e-5,
+                "max_learning_rate": 1e-1,
+            },
+        }
+        ```
+        Here all of the keys are optional. Here are the criterions laid out:
+
+        :key `static_hyperparameters`: These are predefined hyperparameters solely
+        for doing model training and not hyperparameter search. if this key is not
```

Comment:
Are you suggesting that these are hyperparameters that we will use directly for training and not tune via parameter search? If so I think this definition can be improved:
```suggestion
        :key `static_hyperparameters`: These are the subset of hyperparameters that are
        fixed and not tuned via hyperparameter search. if this key is not
```

Reply:
Does this answers this question: https://github.com/neulab/prompt2model/pull/315/files/af6119c63ffb6ce720fa491a415def4c59dc86cf#r1315364617